### PR TITLE
Use the same logic for zip-files as we use for tar-files

### PR DESCRIPTION
**Pull request by lyxell**
Follow-up on #276

### CODE REVIEW

**Comment by lyxell** [comment]
But personally I think this would make more sense:
```bash
tar -tf #{meta[:filename]} | grep -E 'configure|Makefile' | sort | head -1 | xargs dirname
```

**Comment by thedamian** [comment]
Should I trust you lyxell. 

You've never let me down before. This looks good to me but I won't be able to test it till tomorrow

**Comment by minektur** [comment]
I'm fine with the change... but... I'm vaguely opposed to making the codebase more of a mix between bash and ruby.  It makes things harder for potential contributors, having to be passingly fluent in both.

In this case though I dont think the code works.

I have two zip files you can test on - one with everything that unpacks into the current dir and one with a single directory that has two files in it...

http://www.clift.org/fred/Makefile.zip
http://www.clift.org/fred/a.zip

These were my two test cases...

It looks like your code doesn't quite work on those zip files...

In the Makefile case topdir gets set to 

"empty
Makefile"

when it should be set to "." or ""


Unfortunately there are many zip files in the wild that don't have a single directory that contains all the files, and which unpack many files into the current directory.  This is why I add the 'tmpzip' directory - so there is a known place to have everything.

**Comment by lyxell** [comment]
I agree with you @minektur, my main concern is that we at least should handle tars and zips likewise. I'll try your test cases later. They're very appreciated.

**Comment by minektur** [comment]
My only concern is that the requirements are a little different for zip and
tar. Tar archives pretty much universally have a top level containing
directory while zip files work both ways.  Code might be different for
different needs...

**Comment by lyxell** [comment]
@minektur Could you give me a couple of examples of where this applies for zip? I still think there must be a way to handle both the same way and still make it work in all cases.

**Comment by minektur** [comment]
I'll find some examples in the wild.

I agree that there is a way to handle both cases - use the code I submitted
originally :)

**Comment by minektur** [comment]
For the raid controllers I use at work, I get the LSI management utility from here: 

http://www.dzhang.com/blog/2013/03/22/where-to-get-download-lsiutil

grab this file from there

http://www.dzhang.com/files/blog/2013/03/LSIUtil_1.62.zip

and you'll find linux and windows binaries distributed as zip.  That zip file is one of the kind I'm talking about - one that unpacks a bunch of stuff into the current working directory.  It is a real-world, has-bitten-me example of stupid zip packaging of downloadable software.


-----------
 A second example - the managment utility that netgear distributes with it's prosecure UTM firewalls

you can download from here:

https://www.netgear.com/support/product/GS108Ev2.aspx?cid=wmt_netgear_organic  the 'prosafe plus management utility' - ProSafe_Plus_Utility_V2.3.5.zip  - in this case it has only two files - an executable and an html readme that both unpack into the current directory.  This is another real-world example I've used in the course of my day job...


----

You can get the doom source code from gog.com - 

https://www.gog.com/mix/games_with_official_source_code_release

somewhere on that page is doomsrc.zip

which unpacks a readme and two tarballs into the current working directory.



-----


randomly searching found .zip archives from many other sources - western digital, cisco, dell that have multiple files unpacking into the current directory instead of making their own.


Need more examples?


Seriously though, I understand your desire to have consistency in the code.  In this case, the code I wrote is superior to your one-liner in bash for at least two reasons.  The first is that it keeps more bash code from being needlessly added to the code base of a ruby project, and the second is that it handles more cases correctly.  Your objection to my code seems merely to be "it doesn't look the same as the other code".  Am I missing something?

**Comment by lyxell** [comment]
You're not missing something, I'm just not very keen in introducing ~10 extra lines of code for unpacking zip files. None of your latest examples are relevant for Chromebrew though, right? And two of your examples are binary packages which aren't related to the logic we're discussing.

My proposal
```
tar -tf #{meta[:filename]} | grep -E 'configure|Makefile' | sort | head -1 | xargs dirname
```
works in both your original testcases though (a.zip and Makefile.zip).

**Comment by minektur** [comment]
tar -tf #{meta[:filename]} | grep -E


I presume you meant zipinfo -1 .

You probably also want a -i on your grep because you often get 'makefile' instead of 'Makefile' and both work.  You may also want to include other files because you'll likely miss a bunch of zip files that just have a readme telling you to 'cc -o foo foo.c' to build...  or which rely on other 3rd party build systems in newer languages, or which only install scripts, etc.  How will this work in the case of multiple sub-directories that have their own sub-makefiles in them, as is often the case in complex packages?

---------------------------------------

The specific examples I provided are not directly relevant to chromebrew.  The 'zip distribution only' files I personally want to unpack are for binaries that need some mechanism that will let me pick the right binary file to download per arch, and then unpack them.  

http://rclone.org/downloads/

THOSE specific zip files that I care about in this particular instance are packed in a sane way that doesn't unpack into the current directory.  I originally added zip file support to handle those and then realized that binary packages are done differently and that my zip support would only work for source archives.  Then, I dug in and figured out the right thing to do, partly by seeing how homebrew for macos does this same thing for zip archives that they unpack to build.  I personally now have no need for direct zip support in source packages in chromebrew because I need to do something architecture specific.

**I did a little research:**

A brief examination of the homebrew git repo, and a short bash script later yields the following interesting statistics:

Number of homebrew packages that depend on a zipfile in some way:  186 out of 3765 

I downloaded a subset of these (I could easily exclude java related things that have zip source (not talking about jar files... and anything downloaded from github which does the right thing) and examined them.  Of the remaining, I excluded a few that were mac-specific programs or binaries (e.g. platypus which converts bash scripts into macos binaries).   I also manually excluded a few that had .tar.gz sources but used .zip files as data (e.g. openstreetmap data in zip format)

Number of homebrew packages that someone might want some day on chromeos with zip source that unpack into the current directory instead of making their own subdirectory: 21 out of 3765 homebrew packages total.

Most of these are older, or obscure software, and frankly, most people would not miss them.  But someone might want them, and the list I started from is only a sampling of what exists in the world.


The full list, with github links so you can check out the homebrew formulas:

https://github.com/Homebrew/homebrew-core/blob/master/Formula/convertlit.rb
https://github.com/Homebrew/homebrew-core/blob/master/Formula/adplug.rb
https://github.com/Homebrew/homebrew-core/blob/master/Formula/bbcolors.rb
https://github.com/Homebrew/homebrew-core/blob/master/Formula/saxon-b.rb
https://github.com/Homebrew/homebrew-core/blob/master/Formula/saxon.rb
https://github.com/Homebrew/homebrew-core/blob/master/Formula/simh.rb
https://github.com/Homebrew/homebrew-core/blob/master/Formula/mp3gain.rb
https://github.com/Homebrew/homebrew-core/blob/master/Formula/chromedriver.rb
https://github.com/Homebrew/homebrew-core/blob/master/Formula/flasm.rb
https://github.com/Homebrew/homebrew-core/blob/master/Formula/ecm.rb
https://github.com/Homebrew/homebrew-core/blob/master/Formula/zpaq.rb
https://github.com/Homebrew/homebrew-core/blob/master/Formula/unar.rb
https://github.com/Homebrew/homebrew-core/blob/master/Formula/torrentcheck.rb
https://github.com/Homebrew/homebrew-core/blob/master/Formula/osmosis.rb
https://github.com/Homebrew/homebrew-core/blob/master/Formula/gsar.rb
https://github.com/Homebrew/homebrew-core/blob/master/Formula/google-sql-tool.rb
https://github.com/Homebrew/homebrew-core/blob/master/Formula/exomizer.rb
https://github.com/Homebrew/homebrew-core/blob/master/Formula/davmail.rb
https://github.com/Homebrew/homebrew-core/blob/master/Formula/naturaldocs.rb
https://github.com/Homebrew/homebrew-core/blob/master/Formula/locateme.rb
https://github.com/Homebrew/homebrew-core/blob/master/Formula/aap.rb


At this point, I'm done arguing about it.  I think your heuristics are less robust than mine, but that mine does come at a higher line count.  Is the reduction in line count worth giving up the ability to support oddly packaged files?  maybe, maybe not.  

**To summarize my feelings:**

your code:  shorter and more consistent, encourages multi-language mixing, less robust
My code:  more verbose, gets rid of language mixing where possible, more robust


<insert obligatory emerson quote about consistency here :)>

**Comment by lyxell** [comment]
I wouldn't suggest that your solution is more robust.
I tested the packages in the list you provided, and both solutions have the same behaviour on all packages in the list except these cases:

https://github.com/Homebrew/homebrew-core/blob/master/Formula/exomizer.rb
* my solution works: reports './exodecrs'
* your doesn't work: reports '.'

https://github.com/Homebrew/homebrew-core/blob/master/Formula/convertlit.rb
* my solution works: reports './clit18'
* your doesn't work: reports '.'

**Comment by minektur** [comment]
I got an email with your response, but it doesn't seem to be here... anyway, like I said, I'm not likely to be using zip-source functionality soon, so shrug - do whatever...  


------

Discussing your points in the message: For the two pacakges you mentioned in your (missing) message, convertlit.rb - you mention that your soution put you in the clit18 directory and that mine was in error being in . - but there are files in . (above clit18) that need to be used as part of the build - to quote the readme:

    To build under Linux:
            cd lib
            make
            cd ../clit16
            make

I'd say that dropping you in . is more 'right' than dropping you in clit18.


A similar argument could be made for the exomiser package  you  mentioned.  Better to, by default, drop you in the top level directory of what was unpacked, every time.  

Especially if/when other zip packages have a directory structure that has sub-level makefiles, like:

parent/Readme.txt
parent/Makefile
parent/foo/Makefile
parent/foo/stuff
parent/zap/Makefile
parent/zap/stuff


The "... | sort | head -1" will work in this case, but if Makefile were 'makefile' instead, you'd end up in parent/zap instead of parent (because of the sort order...  Are there any zip files like this?  shrug.


The requirements as I see them are:
1) be able to download zip-source-packages
2) cryptographically verify them
3) unpack them
4) have the foo.rb package file's 'build' method called while $CWD is the top-level dir where stuff was unpacked.
5) have the foo.rb package file's 'install' method called while $CWD is the top-level dir where stuff was unpacked

(4 and 5 for you are slightly different I guess because $CWD will be where the config/Makefile is...)

Either your or my code does this in the vast majority of cases, and it doesn't really matter a lot to me.  You've done a lot of good work on this project and while I love to debate and argue.... you have control and have been a thoughtful and responsive committer -  if it's my agreement, or blessing or whatever that you're waiting for, please by all means go ahead, otherwise lets debate some more :)
